### ISSUE TITLE
adding rarefaction (e.g. to transformCounts)

### ISSUE THREAD

**Issue by HenrikEckermann** [microbiome/mia#112]
Hi,

I noticed that mia has already several options to transform counts. I think it would be a good addition to add to the `transformCounts` function to transform counts to rarefied counts. I could have a look at the end of the week if I am able to help with this.

**Comment by antagomir** [comment]
Hi - this is useful but first let us make sure that there are no existing solutions in the (T)SE ecosystem. This is a common task and if we did not implement it for `mia` yet, it may be readily available in the SingleCellExperiment ecosystem. In that case, we could instead 1) add an example to OMA and 2) consider if we need alias of some modified capabilities for microbiome data (which I am not sure of). @FelixErnst or @microsud may have some comments here.

**Comment by antagomir** [comment]
I am not aware of a ready-made implementation. Should be rather straightforward to implement and phyloseq version could provide some insipration.

**Comment by microsud** [comment]
Thanks! Yes I am testing this now using `phyloseq::rarefy_even_depth`. Two things I notice:
1. Returning subsampled assay to `TSE`? 
    If from N samples in `TSE`, 1 sample is removed because it has reads less than user specified **min samples size**, then the output assay has N-1 samples i.e. colnames. Adding this to the `TSE` as new "subsampled" assay returns the TSE with N-1 samples as all assays are subsetted to have same ncol(x). Thus, the original number of samples are not preserved.
2.  Returning subsampled as `altExp` to TSE? 
     In such as case we still have error because it expects the same number of columns.
```
Error in .set_internal_character(x, e, value, getfun = int_colData, setfun = `int_colData<-`,  : 
  invalid 'value' in 'altExp(<TreeSummarizedExperiment>, type="character") <- value':
  'value' should have number of columns equal to 'ncol(x)'
```

The question is: should we use the first option and warn the user that a particular sample is removed from the entire `TSE` and add details in `OMA` or should we return subsampled data inside a `MultiAssayExperiment` container? 

Another thing, this is not the same as transforming the data because rows and columns can change and therefore cannot go with `transformCounts`. I think we need a separate function `getSubsample()`.

**Comment by antagomir** [comment]
This is a broad issue that also touches, for instace, transformations such as `ilr` (not yet implemented) that would return N-1 samples. 

Not clear to me from the description how `getSubsample()` would be used in this case? We might also like to have a mechanism similar to `altExp` (that allows different features but same samples) that would allow different samples but same features. This could be useful for instance in splitting samples by subject in time series (in a (very) roughly similar way that the data can be splitted by ranks to altExps with `splitByRank`). But I do not think that the current `SE`/`MAE` frameworks support such thing.

Using `MAE` could be somewhat problematic because it is primarily intended for different types of complementary data, more than for transformations or subsets. But it is a possible solution.

Perhaps we should just acknowledge than certain operations would lead to subsetting, and those could be considered a more permanent operation once done. Then rarification would just be an operation that permanently changes the dimensionality of the whole (Tree)SE object.

**Comment by antagomir** [comment]
This is also related to issue #128 which is about log-ratio transformations that drop one sample; and to #103 that drops one feature. Sooner or later we would need to converge to some solution at least. The integration of philr to **mia** is pending for this decision, for instance.

**Comment by microsud** [comment]
I think it is more convenient to return modified TreeSE in these cases and let user take care of this. User can for instance place the modified TreeSE within MAE for downstream analyses.

**Comment by antagomir** [comment]
So this would mean that the original assays would be left out from the returned object?

If this is implemented so also SE can be returned if appropriate (same as input).

**Comment by microsud** [comment]
Yes something like this... 
```
library(mia)
library(MultiAssayExperiment)
data("GlobalPatterns")
tse <- GlobalPatterns
tse.subsampled <- subsampleCounts(tse, min_size = 60000, name = "subsampled")
mae <- MultiAssayExperiment::MultiAssayExperiment(c("originalTreeSE" = tse,
                                                    "subsampledTreeSE" = tse.subsampled))
mae
```
Output
```
A MultiAssayExperiment object of 2 listed
 experiments with user-defined names and respective classes.
 Containing an ExperimentList class object of length 2:
 [1] originalTreeSE: TreeSummarizedExperiment with 19216 rows and 26 columns
 [2] subsampledTreeSE: TreeSummarizedExperiment with 12459 rows and 25 columns
Functionality:
 experiments() - obtain the ExperimentList instance
 colData() - the primary/phenotype DataFrame
 sampleMap() - the sample coordination DataFrame
 `$`, `[`, `[[` - extract colData columns, subset, or experiment
 *Format() - convert into a long or wide DataFrame
 assays() - convert ExperimentList to a SimpleList of matrices
 exportClass() - save all data to files
```
The user then for e.g. calculate diversity like this.
```
estimateDiversity(experiments(mae)$subsampledTreeSE, 
                  abund_values = "subsampled",
                  index = "shannon")
```
Gives the user the freedom to use specific `TreeSE` for specific analysis.

**Comment by antagomir** [comment]
One thing to note is that often the low read count samples are permanently discarded in the subsequent analyses. The `MAE` strategy here is possible but I am wondering if it complicates the procedure unnecessarily lot in many cases. The example workflows will have impact on user's choices. A simpler option would be to just discard some samples in the beginning, as a filtering step, then proceed with the plain TreeSE as usual.

**Comment by microsud** [comment]
It is not about just removing samples. Along with samples, there is also removal of taxa after subsampling. These are not the same operations as removing samples with a low number of reads. In which case, only taxa unique to samples that are removed are discarded.

**Comment by FelixErnst** [comment]
Are we sure about this. Rarification is looked upon not quite favourably. Have a look at earlier issues with comments from Amy Willis.

The option with the MAE is good option because it makes only sense to people with more advanced knowledge on Bioc

**Comment by antagomir** [comment]
There is vast discussion about whether or not to rarify. Whereas I advocate searching for alternatives, these also need to be benchmarked against rarified alternatives; and some users may also like to do these comparisons routinely, or, in response to reviewer requests.. Also phyloseq package has decided to include rarification, with additional note that they do not recommend to use it in the analyses. So I would propose that this is added. There can be a note that the concept is arguable.

MAE is indeed perhaps the best solution in this case. There should be some examples on how to use that then.

**Comment by FelixErnst** [comment]
OK. Go for it, but output a warning by default.

**Comment by FelixErnst** [comment]
Closed by #147
